### ISSUE TITLE
[Feature Request] Allow spcifying keys somewhere else

### ISSUE THREAD

**Issue by My1** [BTBurke/caddy-jwt#22]
why do the keys need to be stored in an environment variable? couldnt they also be stored in the caddy file or another file which is kept safely?

also having the keys somewhere else would allow for usage different JSON keys for different paths, for example, different actors allowing access for different paths.

**Comment by BTBurke** [comment]
The main reason is that if you keep your keys in the Caddyfile you can no longer keep it in a public repository.  Putting sensitive information in there is probably not a good idea from a security standpoint either.  Using environment variables for sensitive configuration is a best practice.

The goal of this middleware is the same as for Caddy - to keep things as simple and secure as possible.  You can use the `kid` field in your JWT do your own custom key lookups on the backend and use as many different secrets/algorithms as you like.  Sophisticated security polices are beyond the scope of this middleware, so the right place to do that is in the app that you're proxying to.

**Comment by captncraig** [comment]
Caddy supports substituting environment variables for any part of the caddyfile. See [docs](https://caddyserver.com/docs/caddyfile#env). Is that sufficient for your needs. I keep such things in an .env file and load it when I run with systemd or with [zoo](https://github.com/jsonmaur/zoo).

**Comment by captncraig** [comment]
@BTBurke is it possible to use different keys for different sites within a caddyfile? Perhaps that would be a case where specifying **something** in the caddyfile might be useful.

**Comment by My1** [comment]
@BTBurke well I did specifilically say the caddyfile or another file which is kept safely, also are environment vars that good  of an Idea? I mean especially in linux it is kind of a common practice to let servers and other daemons run as their own user, which cannot log in. but I dont remember a way of setting an environment variable for such a daemon user.

**Comment by BTBurke** [comment]
Sure, in systemd or some other init system you can set an environment variable for the running process.  I run caddy in a container so I haven't tried that, but it shouldn't be a problem.  There are a couple good reasons to go with environment variables:

* Keeps sensitive config outside the Caddyfile
* Allows you to change the secret without Caddy restart
* Not as many corner cases with concurrent read/write access
* Fits in with workflows for 12-factor style apps and containerization

So this assumes that you have some kind of service that is issuing your tokens.  You want that service to be in charge of generating/revoking/rotating your secrets.  Because Caddy doesn't yet have a way to hot reload the config (that I know of), if you put it in the Caddyfile you have no way of having your JWT service and this middleware stay in sync if you change the secret your JWT service is using.  You can't reload the Caddyfile without a restart of the process.  So my personal opinion is that isn't a great way to architect the system, in addition to the fact that I think it's a bad idea to put secrets in the Caddyfile.

You could use a file as a mechanism to pass the secret back and forth but now you are in a situation where you have the potential for concurrent read/write to your secret file at the same time.  You have your JWT service that needs to get an exclusive lock on the file to write a new secret at the same time you may have N goroutines that are trying to read that file to process requests.  You could cache the read at startup, but now you no longer have the ability to change the secret.  Of course, it all depends on which filesystem it's running on and how locks are handled, but I think it could be a source of weird corner cases and bugs.

To @captncraig's point, there is no way to specify multiple keys for different sites in a Caddyfile.  If we were to go this route, one way would be to keep the config in the environment and specify which environment variable contains the key.  The problem with this is that it can potentially open up an attack vector against the JWT in the case of the RS-family of algorithms where an attacker can potentially forge a key by changing the algorithm to HS in the header and signing a token with the public key.  

There are ways to combat this, but it would be by doing stuff that's not in accordance with the JWT spec and/or adding a bunch of extra directives to the middleware and having a more complicated process for validating the config.

So, all that is to say that it quickly becomes complicated.  And in that case, my feeling is that you're treading into the realm of advanced use cases where it makes sense to move your JWT validation into your app where you have full control over your exact set up.

**Comment by My1** [comment]
couldnt it be possible that the key definition could hold the name of the signing algo? that would patch that opening hole right shut again, for example:
`HS256-nKp09d9bMBFPQq8M69v0LVYV16GSkIeJBlk4WDeb`
and of course `RSXXX-pubkey` for RSA with pubkey obviously being the public key.

**Comment by BTBurke** [comment]
There's no hole there.  In fact, if you short-circuit the process of validating the token cryptographically you're giving away information about the signing algorithm, also known as a timing attack.  For example, if you specify HS256 and send a forged token with HS512, but you reject it immediately based on the header, an attacker can measure the response time to figure out what kinds of algorithms you are accepting.

**Comment by My1** [comment]
the hole I referred to was what you said with the RS/HS swap when being able to store the variable somewhere else.
Let me quote you:
> The problem with this is that it can potentially open up an attack vector against the JWT in the case of the RS-family of algorithms where an attacker can potentially forge a key by changing the algorithm to HS in the header and signing a token with the public key.

well this would open a timing to figure out what algorithms are okay, but even then if you literally tell an attacker what algos are accepted, he still has to get in with one of the allowed algorithms. also a small rethoric question: what is safer? 3 doors that can open with the right key or 1 one door which can opened with the right key and the other doors bolted shut? 
putting the fact that all 3 HS algos are pretty much impossible to bruteforce at the speeds we have now
in terms of bruteforcing your way in, having 3 keyholes essentially gives an attacker the option to select the weakest one or to attack all of them at once, while the defender (aka the server) may want to shut anything that is either not used or not secure (HTTPS servers also dont have all possible theoretically secure combinations of algos enabled, usually just the common ones needed by the users, to keep the actual attack surface as small as possible.)

**Comment by BTBurke** [comment]
That's not the right way to think about it.  It's one door with three locks on it and all have the same key.  When people do a brute force attack against an HMAC, they are brute-forcing your key and not the signature.  The security depends on the length and entropy of your secret key, not the algorithm.  It doesn't matter if there are three locks on the door or one.  Once you guess the key, you can open any lock.  That's why taking two locks off the door doesn't do anything.

The bottom line is that you're free to use this library if you want.  You can fork it and put whatever you want in it and build your own version of Caddy from source.  You can move your JWT validation into your app and do what you want with it.  But I'm closing this issue because we're going around and around over nothing.
